backend/src/main/kotlin/co/chainring/core/blockchain/BlockchainTransactionHandler.kt

```
@@ -370,17 +378,37 @@ class BlockchainTransactionHandler(
         logger.debug { "batchHash stored = ${blockchainTx.batchHash} from contract = $onChainBatchHash" }
         if (blockchainTx.batchHash == onChainBatchHash) {
             // batch with this hash is already submitted on onchain (anvil restarted case)
-            completeWithdrawals(WithdrawalEntity.findSettling(chainId))
+            completeWithdrawals(WithdrawalEntity.findSettling(chainId), emptyMap())
         } else {
             // fork case
             blockchainTx.status = BlockchainTransactionStatus.Pending
         }
     }
 
-    private fun completeWithdrawals(settlingWithdrawals: List<WithdrawalEntity>, error: String? = null) {
+    private fun completeWithdrawals(settlingWithdrawals: List<WithdrawalEntity>, withdrawalEvents: Map<Int, Exchange.WithdrawalEventResponse> = emptyMap()) {
         settlingWithdrawals.forEach {
             try {
-                onTxComplete(it.transactionData!!, error)
+                val withdrawAmount = withdrawalEvents[it.sequenceId]?.let { event ->
+                    if (it.amount == BigInteger.ZERO && it.actualAmount != null && event.amount.compareTo(it.actualAmount) != 0) {
+                        logger.warn { "Withdraw all mismatch, sequencer=${it.actualAmount} contract=${event.amount} " }
+                    }
+                    event.amount
+                } ?: it.actualAmount ?: it.amount
```

under what conditions would there not be an event? would it be better to log an error and not complete the withdrawal in that case?

---

yeah there should always be an event. I will adjust this - it was primarily defensive, but better to fail fast if this happens and alert.

---

Actually there is a case where there might not an event albeit rare - if anvil restarts while a withdrawal batch is in progress we do not get a receipt and the code assumes all transactions in the batch were successful - for now I changed to log if we do not get an event, but don't stop or fail anything,